docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,793 @@
+
+By registering a Runtime Extension the Cluster API Runtime became aware of a Runtime Extension implementing a
+Runtime Hook, and as a consequence the runtime starts calling the extension at well-defined moments of the
+Workload Cluster’s lifecycle.
+
+This process has many similarities with registering dynamic webhooks in Kubernetes, but in this case it is required
+to use the following CR:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: RuntimeExtensionConfiguration
+metadata:
+name: "my-amazing-product-runtime-extension"
+webhooks:
+# Name should be fully qualified and unique in the Cluster, thus usage of sub domains, version qualifiers is recommended.
+- name: "my-amazing-runtime-extension.v1.5.panda.com"
+  # List of group/version/hook the RuntimeExtension implements. Required
+  operations:
+  - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"
+    hook: "beforeUpgrade"
+  - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"
+    hook: "afterUpgrade"
+  #ClientConfig defines how to communicate with the RuntimeExtension. Required
+  clientConfig:
+    #`url` gives the location of the RuntimeExtension, in standard URL form (`scheme://host:port/path`). Exactly one of `url` or `service` must be specified.
+    url: "..."
+    service:
+      namespace: "example-namespace"
+      name: "example-service"
+      # `path` is an optional path prefix path which can be sent in any request to this service.
+      path: "runtime-extensions/"
+      # If specified, the port on the service that hosts the RuntimeExtension. Default to 443. `port` should be a valid port number (1-65535, inclusive).
+      port: 8082
+    caBundle: "..."
+  # If specified, define the timeout for each RuntimeExtension call to complete (Default 5s, Max 10s).
+  timeoutSeconds: 2
+  # FailurePolicy defines how errors from RuntimeExtension calls are handled - allowed values are Ignore or Fail. Defaults to Fail.
+  failurePolicy: Fail
+```
+
+Following elements can be added to the above type if consensus is reached/in a follow up iteration:
+
+- Support for defining Runtime Extensions that applies to a subset of Clusters only, e.g.
+```yaml
+# NamespaceSelector decides whether to run the webhook on a Cluster based on whether the namespace for that Cluster matches the selector.
+# If not specified, the WebHook runs for all the namespaces.
+namespaceSelector:
+...
+
+# Or
+
+# ClusterSelector decides whether to run the webhook on a Cluster based on whether the Cluster object matches the selector.
+# If not specified, the WebHook runs for all the Clusters.
+clusterSelector:
+...
+
+# Or a combination of both criteria combined in AND.
+```
+- Support for defining dependencies among Runtime Extensions, e.g. taking inspiration from [systemd unit options](https://www.freedesktop.org/software/systemd/man/systemd.unit.html)
+  Note: designing Runtime Extensions without dependency should always be preferred.
+
```

Comment:
I'd be hesitant to introduce complexity with a dependency system and introduce "pet" like state, unless there's a strong and evident need for it which can be solved with different angle. May be we can consider dropping this for now to avoid confusion?

Reply:
Agreed. created a section with what we don't want to add at this stage/with your considerations, however, I would like to keep it around for tracking that the idea has been discussed